File: internal/workspace/lifecycle.go
package workspace

import (
	"fmt"
	"strings"

	"github.com/ashon/ax/internal/daemon"
)

type LifecycleAction string

const (
	LifecycleActionStart   LifecycleAction = "start"
	LifecycleActionStop    LifecycleAction = "stop"
	LifecycleActionRestart LifecycleAction = "restart"
)

type LifecycleTargetKind string

const (
	LifecycleTargetWorkspace    LifecycleTargetKind = "workspace"
	LifecycleTargetOrchestrator LifecycleTargetKind = "orchestrator"
)

type LifecycleTarget struct {
	Name           string
	Kind           LifecycleTargetKind
	ManagedSession bool
}

type resolvedLifecycleTarget struct {
	target       LifecycleTarget
	workspace    DesiredWorkspace
	orchestrator DesiredOrchestrator
}

// StartNamedTarget starts a configured workspace or managed child orchestrator
// by exact name. Unmanaged targets such as the root orchestrator are rejected.
func StartNamedTarget(socketPath, configPath, target string) (LifecycleTarget, error) {
	return controlNamedTarget(socketPath, configPath, target, LifecycleActionStart)
}

// StopNamedTarget stops a running workspace or managed child orchestrator by
// exact name without deleting its generated artifacts.
func StopNamedTarget(socketPath, configPath, target string) (LifecycleTarget, error) {
	return controlNamedTarget(socketPath, configPath, target, LifecycleActionStop)
}

// RestartNamedTarget recycles a configured workspace or managed child
// orchestrator by exact name. Restart performs a real cleanup/recreate cycle.
func RestartNamedTarget(socketPath, configPath, target string) (LifecycleTarget, error) {
	return controlNamedTarget(socketPath, configPath, target, LifecycleActionRestart)
}

func controlNamedTarget(socketPath, configPath, targetName string, action LifecycleAction) (LifecycleTarget, error) {
	target, err := resolveLifecycleTarget(socketPath, configPath, targetName)
	if err != nil {
		return LifecycleTarget{}, err
	}

	switch target.target.Kind {
	case LifecycleTargetWorkspace:
		if err := controlWorkspaceTarget(socketPath, configPath, target, action); err != nil {
			return LifecycleTarget{}, err
		}
	case LifecycleTargetOrchestrator:
		if err := controlOrchestratorTarget(socketPath, configPath, target, action); err != nil {
			return LifecycleTarget{}, err
		}
	default:
		return LifecycleTarget{}, fmt.Errorf("unsupported target kind %q", target.target.Kind)
	}

	return target.target, nil
}

func resolveLifecycleTarget(socketPath, configPath, targetName string) (resolvedLifecycleTarget, error) {
	targetName = strings.TrimSpace(targetName)
	if targetName == "" {
		return resolvedLifecycleTarget{}, fmt.Errorf("target name is required")
	}

	desired, err := loadDispatchDesiredState(socketPath, configPath)
	if err != nil {
		return resolvedLifecycleTarget{}, err
	}

	workspaceEntry, hasWorkspace := desired.Workspaces[targetName]
	orchestratorEntry, hasOrchestrator := desired.Orchestrators[targetName]

	switch {
	case hasWorkspace && hasOrchestrator:
		return resolvedLifecycleTarget{}, fmt.Errorf("target %q is ambiguous in %s because it matches both a workspace and an orchestrator", targetName, cleanPath(configPath))
	case hasWorkspace:
		return resolvedLifecycleTarget{
			target: LifecycleTarget{
				Name:           workspaceEntry.Name,
				Kind:           LifecycleTargetWorkspace,
				ManagedSession: true,
			},
			workspace: workspaceEntry,
		}, nil
	case hasOrchestrator:
		return resolvedLifecycleTarget{
			target: LifecycleTarget{
				Name:           orchestratorEntry.Name,
				Kind:           LifecycleTargetOrchestrator,
				ManagedSession: orchestratorEntry.ManagedSession,
			},
			orchestrator: orchestratorEntry,
		}, nil
	default:
		return resolvedLifecycleTarget{}, fmt.Errorf("target %q is not defined in %s", targetName, cleanPath(configPath))
	}
}

func controlWorkspaceTarget(socketPath, configPath string, target resolvedLifecycleTarget, action LifecycleAction) error {
	manager := NewManager(socketPath, configPath)

	switch action {
	case LifecycleActionStart:
		if workspaceSessionExists(target.target.Name) {
			return alreadyRunningError(target.target)
		}
		return manager.Create(target.target.Name, target.workspace.Workspace)
	case LifecycleActionStop:
		return stopSessionTarget(target.target)
	case LifecycleActionRestart:
		return manager.Restart(target.target.Name, target.workspace.Workspace)
	default:
		return fmt.Errorf("unsupported lifecycle action %q", action)
	}
}

func controlOrchestratorTarget(socketPath, configPath string, target resolvedLifecycleTarget, action LifecycleAction) error {
	if !target.target.ManagedSession {
		return unsupportedManagedSessionError(target.target, action)
	}

	switch action {
	case LifecycleActionStart:
		if workspaceSessionExists(target.target.Name) {
			return alreadyRunningError(target.target)
		}
		return EnsureOrchestrator(target.orchestrator.Node, target.orchestrator.ParentName, daemon.ExpandSocketPath(socketPath), configPath, true)
	case LifecycleActionStop:
		return stopSessionTarget(target.target)
	case LifecycleActionRestart:
		if err := CleanupOrchestratorState(target.target.Name, target.orchestrator.ArtifactDir); err != nil {
			return err
		}
		return EnsureOrchestrator(target.orchestrator.Node, target.orchestrator.ParentName, daemon.ExpandSocketPath(socketPath), configPath, true)
	default:
		return fmt.Errorf("unsupported lifecycle action %q", action)
	}
}

func stopSessionTarget(target LifecycleTarget) error {
	if !workspaceSessionExists(target.Name) {
		return notRunningError(target)
	}
	if err := workspaceDestroySession(target.Name); err != nil {
		return fmt.Errorf("destroy tmux session: %w", err)
	}
	return nil
}

func alreadyRunningError(target LifecycleTarget) error {
	return fmt.Errorf("%s %q is already running", target.Kind, target.Name)
}

func notRunningError(target LifecycleTarget) error {
	return fmt.Errorf("%s %q is not running", target.Kind, target.Name)
}

func unsupportedManagedSessionError(target LifecycleTarget, action LifecycleAction) error {
	return fmt.Errorf("%s %q does not support targeted %s because it is not a managed session", target.Kind, target.Name, action)
}

File: rs/crates/ax-workspace/src/lifecycle.rs
//! Targeted lifecycle control for managed sessions.
//!
//! Mirrors `internal/workspace/lifecycle.go`.

use std::path::Path;
use std::path::{Component, PathBuf};

use ax_proto::types::{LifecycleAction, LifecycleTarget, LifecycleTargetKind};

use crate::{
    cleanup_orchestrator_state, ensure_orchestrator, load_dispatch_desired_state,
    DesiredOrchestrator, DesiredWorkspace, Manager, TmuxBackend,
};

#[derive(Debug, thiserror::Error)]
pub enum LifecycleError {
    #[error("target name is required")]
    TargetRequired,
    #[error(transparent)]
    Dispatch(#[from] crate::DispatchError),
    #[error("target {target:?} is ambiguous in {config} because it matches both a workspace and an orchestrator")]
    AmbiguousTarget { target: String, config: String },
    #[error("target {target:?} is not defined in {config}")]
    TargetNotDefined { target: String, config: String },
    #[error("{kind} {name:?} is already running")]
    AlreadyRunning { kind: &'static str, name: String },
    #[error("{kind} {name:?} is not running")]
    NotRunning { kind: &'static str, name: String },
    #[error(
        "{kind} {name:?} does not support targeted {action} because it is not a managed session"
    )]
    UnsupportedManagedSession {
        kind: &'static str,
        name: String,
        action: &'static str,
    },
    #[error("destroy tmux session: {0}")]
    DestroySession(#[source] ax_tmux::TmuxError),
    #[error(transparent)]
    Workspace(#[from] crate::WorkspaceError),
    #[error(transparent)]
    Orchestrator(#[from] crate::OrchestratorError),
}

#[derive(Debug, Clone)]
struct ResolvedLifecycleTarget {
    target: LifecycleTarget,
    workspace: Option<DesiredWorkspace>,
    orchestrator: Option<DesiredOrchestrator>,
}

pub fn start_named_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target: &str,
) -> Result<LifecycleTarget, LifecycleError> {
    control_named_target(
        tmux,
        socket_path,
        config_path,
        ax_bin,
        target,
        LifecycleAction::Start,
    )
}

pub fn stop_named_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target: &str,
) -> Result<LifecycleTarget, LifecycleError> {
    control_named_target(
        tmux,
        socket_path,
        config_path,
        ax_bin,
        target,
        LifecycleAction::Stop,
    )
}

pub fn restart_named_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target: &str,
) -> Result<LifecycleTarget, LifecycleError> {
    control_named_target(
        tmux,
        socket_path,
        config_path,
        ax_bin,
        target,
        LifecycleAction::Restart,
    )
}

fn control_named_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target_name: &str,
    action: LifecycleAction,
) -> Result<LifecycleTarget, LifecycleError> {
    let target = resolve_lifecycle_target(socket_path, config_path, target_name)?;
    match &target.target.kind {
        LifecycleTargetKind::Workspace => {
            control_workspace_target(tmux, socket_path, config_path, ax_bin, &target, action)?
        }
        LifecycleTargetKind::Orchestrator => {
            control_orchestrator_target(tmux, socket_path, config_path, ax_bin, &target, action)?
        }
    }
    Ok(target.target)
}

fn resolve_lifecycle_target(
    socket_path: &Path,
    config_path: &Path,
    target_name: &str,
) -> Result<ResolvedLifecycleTarget, LifecycleError> {
    let target_name = target_name.trim();
    if target_name.is_empty() {
        return Err(LifecycleError::TargetRequired);
    }

    let desired = load_dispatch_desired_state(socket_path, config_path)?;
    let workspace = desired.workspaces.get(target_name).cloned();
    let orchestrator = desired.orchestrators.get(target_name).cloned();

    match (workspace, orchestrator) {
        (Some(_), Some(_)) => Err(LifecycleError::AmbiguousTarget {
            target: target_name.to_owned(),
            config: clean_path(&config_path.display().to_string()),
        }),
        (Some(workspace), None) => Ok(ResolvedLifecycleTarget {
            target: LifecycleTarget {
                name: workspace.name.clone(),
                kind: LifecycleTargetKind::Workspace,
                managed_session: true,
            },
            workspace: Some(workspace),
            orchestrator: None,
        }),
        (None, Some(orchestrator)) => Ok(ResolvedLifecycleTarget {
            target: LifecycleTarget {
                name: orchestrator.name.clone(),
                kind: LifecycleTargetKind::Orchestrator,
                managed_session: orchestrator.managed_session,
            },
            workspace: None,
            orchestrator: Some(orchestrator),
        }),
        (None, None) => Err(LifecycleError::TargetNotDefined {
            target: target_name.to_owned(),
            config: clean_path(&config_path.display().to_string()),
        }),
    }
}

fn control_workspace_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target: &ResolvedLifecycleTarget,
    action: LifecycleAction,
) -> Result<(), LifecycleError> {
    let workspace = target.workspace.as_ref().expect("workspace target");
    let manager = Manager::with_tmux(
        socket_path.to_path_buf(),
        Some(config_path.to_path_buf()),
        ax_bin.to_path_buf(),
        tmux.clone(),
    );

    match action {
        LifecycleAction::Start => {
            if tmux.session_exists(&target.target.name) {
                return Err(LifecycleError::AlreadyRunning {
                    kind: "workspace",
                    name: target.target.name.clone(),
                });
            }
            manager.create(&target.target.name, &workspace.workspace)?;
            Ok(())
        }
        LifecycleAction::Stop => stop_session_target(tmux, &target.target),
        LifecycleAction::Restart => {
            manager.restart(&target.target.name, &workspace.workspace)?;
            Ok(())
        }
    }
}

fn control_orchestrator_target<B: TmuxBackend + Clone>(
    tmux: &B,
    socket_path: &Path,
    config_path: &Path,
    ax_bin: &Path,
    target: &ResolvedLifecycleTarget,
    action: LifecycleAction,
) -> Result<(), LifecycleError> {
    let orchestrator = target.orchestrator.as_ref().expect("orchestrator target");
    if !target.target.managed_session {
        return Err(LifecycleError::UnsupportedManagedSession {
            kind: "orchestrator",
            name: target.target.name.clone(),
            action: lifecycle_action_name(&action),
        });
    }

    match action {
        LifecycleAction::Start => {
            if tmux.session_exists(&target.target.name) {
                return Err(LifecycleError::AlreadyRunning {
                    kind: "orchestrator",
                    name: target.target.name.clone(),
                });
            }
            ensure_orchestrator(
                tmux,
                &orchestrator.node,
                &orchestrator.parent_name,
                socket_path,
                Some(config_path),
                ax_bin,
                true,
            )?;
            Ok(())
        }
        LifecycleAction::Stop => stop_session_target(tmux, &target.target),
        LifecycleAction::Restart => {
            cleanup_orchestrator_state(tmux, &target.target.name, &orchestrator.artifact_dir)?;
            ensure_orchestrator(
                tmux,
                &orchestrator.node,
                &orchestrator.parent_name,
                socket_path,
                Some(config_path),
                ax_bin,
                true,
            )?;
            Ok(())
        }
    }
}

fn stop_session_target<B: TmuxBackend>(
    tmux: &B,
    target: &LifecycleTarget,
) -> Result<(), LifecycleError> {
    if !tmux.session_exists(&target.name) {
        return Err(LifecycleError::NotRunning {
            kind: lifecycle_kind_name(&target.kind),
            name: target.name.clone(),
        });
    }
    tmux.destroy_session(&target.name)
        .map_err(LifecycleError::DestroySession)?;
    Ok(())
}

fn lifecycle_kind_name(kind: &LifecycleTargetKind) -> &'static str {
    match kind {
        LifecycleTargetKind::Workspace => "workspace",
        LifecycleTargetKind::Orchestrator => "orchestrator",
    }
}

fn lifecycle_action_name(action: &LifecycleAction) -> &'static str {
    match action {
        LifecycleAction::Start => "start",
        LifecycleAction::Stop => "stop",
        LifecycleAction::Restart => "restart",
    }
}

fn clean_path(path: &str) -> String {
    if path.trim().is_empty() {
        return String::new();
    }
    normalize_path(Path::new(path)).display().to_string()
}

fn normalize_path(path: &Path) -> PathBuf {
    let mut out = PathBuf::new();
    let is_absolute = path.is_absolute();

    for component in path.components() {
        match component {
            Component::CurDir => {}
            Component::ParentDir => {
                if !out.pop() && !is_absolute {
                    out.push("..");
                }
            }
            Component::RootDir | Component::Prefix(_) | Component::Normal(_) => {
                out.push(component.as_os_str());
            }
        }
    }

    if out.as_os_str().is_empty() {
        if is_absolute {
            PathBuf::from("/")
        } else {
            PathBuf::from(".")
        }
    } else {
        out
    }
}
